modelator/pytest/decorators.py

```diff
@@ -0,0 +1,117 @@
+import json
+from inspect import signature
+from typing import Callable
+
+import pytest
+from pytest import FixtureRequest
+
+from modelator.Model import Model
+
+
+def dict_get_keypath(data, property_path=None):
+    if property_path is not None:
+        keys = property_path.split(".")
+        for key in keys:
+            match data:
+                case dict():
+                    data = data[key]
+                case list():
+                    data = data[int(key)]
+    return data
+
+
+def get_args(func: Callable):
+    params = signature(func).parameters.values()
+    return [param.name for param in params if param.kind == param.POSITIONAL_OR_KEYWORD]
+
+
+def itf(filepath: str, keypath=None):
+    def decorator(func: Callable) -> Callable:
+        with open(filepath, encoding="utf-8") as f:
+            trace = json.load(f)["states"]
+
+        if keypath is None:
+            step_fixtures = ["mbt_step"]
+        else:
+            step_fixtures = [f"mbt_{dict_get_keypath(step, keypath)}" for step in trace]
+
+        @pytest.mark.usefixtures(*step_fixtures)
+        def wrapper(request: FixtureRequest):
+            for step in trace:
+                try:
+                    if keypath is None:
+                        step_func = request.getfixturevalue("mbt_step")
+                    else:
+                        step_func = request.getfixturevalue(
+                            f"mbt_{dict_get_keypath(step, keypath)}"
+                        )
+                    kwargs = {
+                        arg: step[arg] if arg in step else request.getfixturevalue(arg)
+                        for arg in get_args(step_func)
+                    }
+                    step_func(**kwargs)
+                except pytest.FixtureLookupError as ex:
+                    raise RuntimeError("fixture not found") from ex
+            kwargs = {arg: request.getfixturevalue(arg) for arg in get_args(func)}
+            func(**kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+def mbt(tlapath: str, keypath=None, **kwargs):
```

Comment:
what is the `keypath` argument here? Is there an example among tests?

Reply:
I will add a test.

`keypath` is supposed to be the path, a sequence of nested keys, of the value that you want to match with _tag_ in `@step(tag)`.

Reply:
Added a new test. See,
- `tests/test_pytest_bingame.py`
- `tests/models/bingame.tla`